planetiler-core/src/main/java/com/onthegomap/planetiler/geo/GeoUtils.java

```diff
@@ -429,37 +432,47 @@ public static boolean isConvex(LinearRing ring) {
     double dx1 = c1x - c0x;
     double dy1 = c1y - c0y;
 
-    int sign = 0;
+    double negZ = 1e-20, posZ = 1e-20;
 
-    for (; i < seq.size(); i++) {
-      double c2x = seq.getX(i);
-      double c2y = seq.getY(i);
+    // need to wrap around to make sure the triangle formed by last and first points does not change sign
+    for (; i <= size + 1; i++) {
+      // first and last point should be the same, so skip index 0
+      int idx = i < size ? i : (i + 1 - size);
+      double c2x = seq.getX(idx);
+      double c2y = seq.getY(idx);
 
       double dx2 = c2x - c1x;
       double dy2 = c2y - c1y;
       double z = dx1 * dy2 - dy1 * dx2;
 
-      // if z == 0 (with small delta to account for rounding errors) then keep skipping
-      // points to ignore repeated or collinear points
-      if (Math.abs(z) < 1e-10) {
-        continue;
+      double absZ = Math.abs(z);
+
+      // look for sign changes in the triangles formed by sequential points
+      // but, we want to allow for rounding errors and small concavities relative to the overall shape
+      // so track the largest positive and negative threshold for triangle area and compare them once we
+      // have enough points
+      boolean check = i >= minPointsToCheck;
+      if (z < 0 && absZ > negZ) {
+        negZ = absZ;
+      } else if (z > 0 && absZ > posZ) {
+        posZ = absZ;
+      } else if (i > minPointsToCheck) { // always check at i=minPointsToCheck
```

Comment:
Since check isn't being used before being reassigned here, couldn't the declaration be simplified to `boolean check = i == minPointsToCheck`? Doesn't seem like intended behavior

Reply:
Once we've seen over the minimum number of points, we do want to check any time we extend `negZ` or `posZ` - in that case we'll hit the first or second case and not this one.

Reply:
Update makes this much clearer 👍